### ISSUE TITLE
Alternatives to Count Multivariate Metrics approach

### ISSUE THREAD

**Issue by allenmichael099** [e-mission/e-mission-docs#1032]

**Comment by allenmichael099** [comment]
As an example, the true distribution (rv1) on 3 modes could be [0.9, .05, .05], but with the confusion matrix dataset, we got [0.99, 0.005, .005] (rv2). The difference between these, on average, will be more than the estimated variance from rv2.

Each row in the table below represents a mode. The error found is outside 2 variances for each mode.

``` 
NMC = 10000
orig_n = 1000

rv1 = pd.DataFrame(scipy.stats.multinomial.rvs(orig_n, p = [0.9, .05, .05], size= NMC, random_state=None))
rv2 = pd.DataFrame(scipy.stats.multinomial.rvs(orig_n, p = [0.99, 0.005, .005], size= NMC, random_state=None))

print(pd.DataFrame({
    "mean 1":rv1.mean(axis=0),
    "mean 2":rv2.mean(axis=0),
    "diff":rv1.mean(axis=0) - rv2.mean(axis=0),
    "var 2":rv2.var(axis=0),
    "sd 2":np.sqrt(rv2.var(axis=0)),
}).round(3).to_markdown())
```
|    |   mean 1 |   mean 2 |    diff |   var 2 |   sd 2 |
|---:|---------:|---------:|--------:|--------:|-------:|
|  0 |  900.045 |  989.994 | -89.949 |   9.837 |  3.136 |
|  1 |   50.083 |    5.017 |  45.066 |   4.986 |  2.233 |
|  2 |   49.872 |    4.989 |  44.884 |   4.909 |  2.216 |

**Comment by allenmichael099** [comment]
This is more of a problem for modes with high or low estimated proportions. For example, if
rv1 ~ Multinom(1000, [0.8, .05, .15]), and 
rv2 ~ Multinom(1000, [0.5, .35, .15]), 
then the difference on average between rv1 and rv2's means will be just above 1 variance from rv2, as in the table below. However, if you swap their roles, so in the confusion matrix dataset you found rv1's distribution but in the real world you found rv2's distribution, and you used the rv1 estimated variance, you would be off by several variances for row 1's mode.

The estimated variance for a single mode in this setting will be bounded above by $n \cdot 0.5^2$,
and will shrink as p gets further from 0.5. So the variance will perform better for modes that we estimate to be more prominent.
|    |   mean 1 |   mean 2 |     diff |   var 1 |   var 2 |   sd 2 |
|---:|---------:|---------:|---------:|--------:|--------:|-------:|
|  0 |  800.08  |  500.368 |  299.712 | 162.333 | 250.654 | 15.832 |
|  1 |   49.895 |  349.724 | -299.829 |  48.699 | 231.889 | 15.228 |
|  2 |  150.025 |  149.908 |    0.118 | 129.832 | 127.362 | 11.285 |

**Comment by allenmichael099** [comment]
As Shankari writes:
"With the multinomial method for estimating uncertainty (as outlined in Grace's paper [Count Multivariate Metrics]), the quality of the model (70% accurate or 90% accurate) does not matter wrt the technique for computing the error bars. The error bars may be smaller in one case than the other, but we still need to use the variance, no standard deviation. 
Improving model accuracy will not fix the usage of variance."

One alternative that requires a better classifier to work:
1. estimate the matrix P = P(predicted | true) from the confusion matrix. This depends less on the distribution of modes in the confusion matrix dataset. I transpose this if the true modes are on the rows.
2. Calculate r = vector of proportions of mode predictions. 
3. true mode proportions t $\approx$ $(P^T)^{-1} \cdot r$. (I used Pseudoinverse thinking it would work better, but that might not matter. Avoid directly inverting the matrix, since that leads to numerical errors.)

This come from solving a system of equations relating the proportions of predictions to the true mode proportions. $\pi_j$ below represents true mode proportion for mode j.
![image](https://github.com/e-mission/e-mission-docs/assets/56986329/1002b7f8-840a-45e5-97a2-7fdfa19429da)

**Comment by allenmichael099** [comment]
Two examples from a simulation: 
confusion matrix and unlabeled datasets have 10,000 trips each.
The model predicts each mode based on the true mode with the probabilities specified in the rows below:
<img width="182" alt="image" src="https://github.com/e-mission/e-mission-docs/assets/56986329/c0c304d2-d627-4e21-a8e7-2620dea614ee">

The estimated probabilities are:
<img width="222" alt="image" src="https://github.com/e-mission/e-mission-docs/assets/56986329/6ca58d60-bc4e-435e-8124-2b7ea78a38b5">

Calculate estimates with both methods. Each uses the predictions from the unlabeled dataset and the confusion matrix.

**If both datasets have the mode distribution of a, b, c, and d with [40, 30, 20, 10]%:**
multinomial method:[40.10, 29.93, 20.06, 9.91]% The estimates were within one standard deviation.
alternative method: [40.17, 30.02, 20.08, 9.73]. Slightly worse.

**What happens if the unlabeled dataset has a distribution of [10, 20, 30, 40]%?**
Predictions: [11.56, 30.23, 26.51, 31.70]%
multinomial method: [24.49, 33.83, 22.78, 18.90]%. 
The counts are not within 2 sd but are within 2 variances. The predictions did better than the multinomial method in this case. Using this could lead us to mistakenly believe that mode 'b' is the most used mode, when 'd' is actually the most used.

alternative method: [9.41, 18.50, 30.67, 41.42]%. 
The true proportions were estimated fairly well- we get the correct ordering by mode prominence. I just don't know what the appropriate uncertainty for this estimate is.

**Comment by allenmichael099** [comment]
This alternative method doesn't work nicely when the classifier does not predict mode i for true mode i >= 50% of the time.
For example, when I build a confusion matrix for the sensing algorithm based on Community Cycles (cc) data and make estimates on the rest of the CEO data, I get negative values and a bus proportion greater than 1, since those happen to solve the system of equations.  The rows of the top table below shows P(predicted | true) for cc data. The middle table shows true and estimated proportions of each mode present in non cc data. The bottom table's columns show P(true | predicted) on cc data. The modes predictions that would need to improve for the alternative method to work are bike/ebike, bus, and train.

![image](https://github.com/e-mission/e-mission-docs/assets/56986329/06488b08-9151-49c7-b1e8-06b7d6e593ee)
![image](https://github.com/e-mission/e-mission-docs/assets/56986329/19405734-6064-4355-ae3c-a3d73b0134bf)

<img width="371" alt="image" src="https://github.com/e-mission/e-mission-docs/assets/56986329/ac539933-5acc-4dcd-895f-f343d0f9770a">

**Comment by shankari** [comment]
wrt https://github.com/e-mission/e-mission-docs/issues/1032#issuecomment-1821669412
I will point out that it is more likely that error falls outside the error bars as the model is better.
Basically, with the better model, the uncertainty is lower, and so even with small errors, they are outside the error bars.

You can re-try the "classic method" with a poor model (e.g. highest probability is 0.6 and 0.66) and I will bet that the error will consistently be within the variance,.

**Comment by shankari** [comment]
@allenmichael099 interesting. IFF you can figure out how to estimate the uncertainty for the alternative method, I think it might be worthwhile exploring an ensemble of the "classic" and "alternative" methods, depending on the prediction percentages in the confusion matrix.

**Comment by allenmichael099** [comment]
@shankari I agree that with the classic method, the poor model with probabilities near 0.6 will be more likely to have the error within one variance. I think it will still matter how far apart the labeled and unlabeled  true mode distributions are, and can check.

I'm not sure at the moment how to ensemble the two beyond selecting which to use depending on how good the classifier is. I still don't know how to find the uncertainty of the alternative method.

**Comment by shankari** [comment]
I just want to make sure that we are not currently hung up or focused on the quality of the model because the classic method works better for our use case (medium-quality model) in terms of getting a value which is within the error bars.
I propose that we implement the classic method with the current model for now since it is strictly better than what we have now, which is no error bars, and table this issue until we have more time to focus on it.
